File: src/services/entity-intelligence/contracts.ts
// TypeScript mirror of `EntityIntelligenceEnvelope` DTO.
//
// Parity contract for the Rust producer at
// `src-tauri/abilities-runtime/src/abilities/get_entity_intelligence/contracts.rs`.
// Renames Rust snake_case fields to camelCase per the producer's serde annotations.
// Per L0 packet §5.1 and locked decisions in §13 (server-signed cursor pagination,
// per-fact `ProvenanceRef`, typed empty reasons, 1-outer-N-inner block model).

export type EntityKind = "account" | "project" | "person" | "meeting";

export type ContextDepth = "shallow" | "standard" | "deep";

export type EnvelopeSectionV1 =
  | "facts"
  | "health"
  | "metadata_proposals"
  | "open_loops"
  | "touchpoints"
  | "threads"
  | "record";

export type EnvelopeSectionV2 = EnvelopeSectionV1 | "relationships";

export type EnvelopeSection = EnvelopeSectionV2;

export const ENVELOPE_SCHEMA_VERSION_V1 = 1;
export const ENVELOPE_SCHEMA_VERSION_V2 = 2;
export const ENVELOPE_SCHEMA_VERSION = ENVELOPE_SCHEMA_VERSION_V2;

export type EnvelopeSchemaVersion =
  | typeof ENVELOPE_SCHEMA_VERSION_V1
  | typeof ENVELOPE_SCHEMA_VERSION_V2;

export type EnvelopeSectionForSchema<Version extends EnvelopeSchemaVersion> =
  Version extends typeof ENVELOPE_SCHEMA_VERSION_V1
    ? EnvelopeSectionV1
    : EnvelopeSectionV2;

interface EntityIntelligenceInputBase {
  entityType: EntityKind;
  entityId: string;
  depth: ContextDepth;
}

export type EntityIntelligenceInputV1 = EntityIntelligenceInputBase & {
  schemaVersion: typeof ENVELOPE_SCHEMA_VERSION_V1;
  sections?: EnvelopeSectionV1[];
};

export type EntityIntelligenceInputV2 = EntityIntelligenceInputBase & {
  schemaVersion: typeof ENVELOPE_SCHEMA_VERSION_V2;
  sections?: EnvelopeSectionV2[];
};

export type EntityIntelligenceInput =
  | EntityIntelligenceInputV1
  | EntityIntelligenceInputV2;

// ---- subject ---------------------------------------------------------------

export type SubjectRef =
  | { account: string }
  | { project: string }
  | { person: string }
  | { meeting: string }
  | { user: string }
  | "global"
  | { multi: SubjectRef[] }
  | "unknown";

export interface NormalizedSubject {
  kind: EntityKind;
  id: string;
  subjectRef: SubjectRef;
  displayLabel: string;
}

// ---- cursor + pagination ---------------------------------------------------

export type Cursor = string;

export type CursorState =
  | { kind: "stable" }
  | { kind: "data_shifted"; advisory: string }
  | { kind: "invalidated"; reason: string; restart_required: boolean };

export interface Paginated<T> {
  items: T[];
  nextCursor: Cursor | null;
  totalHint: number | null;
  cursorState: CursorState;
}

// ---- empty + section state ------------------------------------------------

export type EmptyReasonV1 =
  | "not_connected"
  | "not_processed_yet"
  | "filtered_out_by_subject"
  | "no_relevant_touchpoints"
  | "stale"
  | "no_evidence_backed_proposal"
  | "unsupported_for_subject"
  | "not_requested"
  | { partial_failure: { advisory: string } };

export type EmptyReasonV2 = EmptyReasonV1 | "no_relevant_relationships";

export type EmptyReason = EmptyReasonV2;

export type EmptyReasonForSchema<Version extends EnvelopeSchemaVersion> =
  Version extends typeof ENVELOPE_SCHEMA_VERSION_V1 ? EmptyReasonV1 : EmptyReasonV2;

export type SectionStateV1 =
  | { kind: "present"; item_count: number }
  | { kind: "empty"; reason: EmptyReasonV1 };

export type SectionStateV2 =
  | { kind: "present"; item_count: number }
  | { kind: "empty"; reason: EmptyReasonV2 };

export type SectionState = SectionStateV2;

export type SectionStateForSchema<Version extends EnvelopeSchemaVersion> =
  Version extends typeof ENVELOPE_SCHEMA_VERSION_V1 ? SectionStateV1 : SectionStateV2;

// ---- per-fact provenance reference (ADR-0130 §2 amendment) ----------------

export interface ProvenanceRef {
  sourceIds: string[];
}

// ---- envelope-level provenance + trust + sensitivity ----------------------

export interface EnvelopeProvenanceSource {
  id: string;
  label: string;
  sourceType: string | null;
  asOf: string | null;
  redacted: boolean;
}

export interface EnvelopeProvenance {
  sources: EnvelopeProvenanceSource[];
  redactionApplied: boolean;
}

export type TrustBand =
  | "likely_current"
  | "use_with_caution"
  | "needs_verification"
  | "unscored";

export interface EnvelopeTrustSummary<Section extends EnvelopeSection = EnvelopeSection> {
  aggregateBand: TrustBand;
  sectionCaveats: Partial<Record<Section, string>>;
}

export type ClaimSensitivity =
  | "public"
  | "internal"
  | "confidential"
  | "user_only";

// ---- claim-backed primitives ----------------------------------------------

export type Freshness = "current" | "aging" | "stale" | "unknown";

export type ClaimState = "active" | "dormant" | "tombstoned" | "withdrawn";

export type SurfacingState = "active" | "dormant";

export type ClaimVerificationState =
  | "active"
  | "contested"
  | "needs_user_decision";

export type RenderSurface =
  | "tauri_entity_detail"
  | "tauri_briefing_prep"
  | "tauri_meeting_detail"
  | "tauri_email_summary"
  | "action"
  | "tauri_provenance"
  | "tauri_report"
  | "tauri_chat"
  | "mcp_tool"
  | "mcp_tool_detail"
  | "p2_publication"
  | "log_structured"
  | "push_notification";

export type RenderPolicyKind = "render" | "redacted" | "drop";

export type RedactionAffordance =
  | {
      kind: "confidential_click_to_reveal";
      claim_id: string;
      label: string;
      audit_required: boolean;
    }
  | { kind: "confidential_hidden"; label: string }
  | { kind: "user_only_hidden"; label: string };

export interface RenderPolicy {
  kind: RenderPolicyKind;
  sensitivity: ClaimSensitivity;
  surface: RenderSurface;
  claimId?: string | null;
  affordance?: RedactionAffordance | null;
}

export interface RenderableClaimText {
  text: string;
  policy: RenderPolicy;
}

export interface EntityFact {
  claimId: string;
  subjectRef: SubjectRef;
  fieldPath: string | null;
  claimType: string;
  renderedText: RenderableClaimText;
  trustBand: TrustBand;
  freshness: Freshness;
  sourceAsof: string | null;
  sensitivity: ClaimSensitivity;
  lifecycleState: ClaimState;
  surfacingState: SurfacingState;
  verificationState: ClaimVerificationState;
  provenance: ProvenanceRef;
}

export interface HealthStoryRow {
  label: string;
  body: string;
  evidenceClaimIds: string[];
  provenance: ProvenanceRef;
}

export interface HealthStory {
  headline: string | null;
  rows: HealthStoryRow[];
}

export interface MetadataProposal {
  proposalId: string;
  subjectRef: SubjectRef;
  fieldPath: string;
  currentValue: string | null;
  proposedValue: string;
  trustBand: TrustBand;
  sensitivity: ClaimSensitivity;
  provenance: ProvenanceRef;
}

// ---- open loop with receipt ----------------------------------------------

export interface ReceiptTargetRef {
  claimId: string;
  subjectRef: SubjectRef;
  fieldPath: string | null;
}

export interface OpenLoopSubject {
  entity_type: string;
  entity_id: string;
}

export interface OpenLoop {
  id: string;
  subject: OpenLoopSubject;
  loop_kind: string;
  description: string;
  owner?: string | null;
  due_date?: string | null;
  status?: string | null;
  source_asof?: string | null;
  claim_type: string;
}

export interface OpenLoopWithReceipt {
  openLoop: OpenLoop;
  receiptTarget: ReceiptTargetRef;
  trustBand: TrustBand;
  freshness: Freshness;
  provenance: ProvenanceRef;
}

// ---- touchpoint -----------------------------------------------------------

export type TouchpointKind =
  | "meeting"
  | "email_thread"
  | "document"
  | "salesforce"
  | "linear";

export type InclusionReason =
  | "subject_match"
  | "entity_link"
  | "attendee_match"
  | "domain_match";

export type ExclusionReason =
  | "subject_mismatch"
  | "outside_window"
  | "low_confidence"
  | "suppressed";

export interface Touchpoint {
  meetingId: string | null;
  kind: TouchpointKind;
  when: string;
  subjectRef: SubjectRef;
  inclusionReason: InclusionReason;
  exclusionReason: ExclusionReason | null;
  trustBand: TrustBand;
  freshness: Freshness;
  provenance: ProvenanceRef;
}

export interface CandidateSetRef {
  windowStart: string | null;
  windowEnd: string | null;
  filterDescription: string;
}

export interface SubjectScope {
  primary: SubjectRef;
  alsoIncludes: SubjectRef[];
}

export interface TouchpointBundle<Reason extends EmptyReason = EmptyReason> {
  upcoming: Paginated<Touchpoint>;
  recent: Paginated<Touchpoint>;
  candidateSet: CandidateSetRef;
  emptyReason: Reason | null;
  subjectScope: SubjectScope;
}

// ---- relationships + participation ---------------------------------------

export type RelationshipInclusionReason =
  | "subject_match"
  | "hierarchy"
  | "explicit_link"
  | "attendee_match"
  | "co_attendance"
  | "work_item"
  | "content_link";

export interface RelationshipEdge {
  edgeId: string;
  edgeType: string;
  subjectRef: SubjectRef;
  relatedSubjectRef: SubjectRef;
  relatedDisplayLabel: RenderableClaimText | null;
  observedAt: string | null;
  sourceAsof: string | null;
  confidence: number;
  sensitivity: ClaimSensitivity;
  inclusionReason: RelationshipInclusionReason;
  traversalDepth: number;
  trustBand: TrustBand;
  freshness: Freshness;
  provenance: ProvenanceRef;
  caveats: string[];
}

export interface RelationshipParticipant {
  subjectRef: SubjectRef;
  displayLabel: RenderableClaimText | null;
  role: RenderableClaimText | null;
  relationship: RenderableClaimText | null;
  sensitivity: ClaimSensitivity;
  normalizedTouchpointCount: number;
  recentTouchpointIds: string[];
  lastSeenAt: string | null;
  trustBand: TrustBand;
  freshness: Freshness;
  provenance: ProvenanceRef;
  caveats: string[];
}

export interface RelationshipTruncation {
  edgesTruncated: boolean;
  participantsTruncated: boolean;
  perEdgeCap: number;
}

export interface RelationshipsBundle {
  edges: Paginated<RelationshipEdge>;
  participants: Paginated<RelationshipParticipant>;
  candidateSet: CandidateSetRef;
  emptyReason: EmptyReasonV2 | null;
  subjectScope: SubjectScope;
  truncation: RelationshipTruncation;
  caveats: string[];
}

// ---- threads + record ----------------------------------------------------

export interface ThreadSummary {
  threadId: string;
  title: string | null;
  lastActivityAt: string | null;
  messageCount: number;
  provenance: ProvenanceRef;
}

export interface RecordEntry {
  claimId: string;
  subjectRef: SubjectRef;
  claimType: string;
  recordedAt: string;
  renderedText: RenderableClaimText;
  trustBand: TrustBand;
  sensitivity: ClaimSensitivity;
  provenance: ProvenanceRef;
}

// ---- envelope --------------------------------------------------------------

interface EntityIntelligenceEnvelopeBase<Version extends EnvelopeSchemaVersion> {
  subject: NormalizedSubject;
  sections: Partial<
    Record<EnvelopeSectionForSchema<Version>, SectionStateForSchema<Version>>
  >;
  facts: Paginated<EntityFact>;
  healthStory: HealthStory | null;
  metadataProposals: Paginated<MetadataProposal>;
  openLoops: Paginated<OpenLoopWithReceipt>;
  touchpoints: Paginated<TouchpointBundle<EmptyReasonForSchema<Version>>>;
  threads: Paginated<ThreadSummary>;
  recordEntries: Paginated<RecordEntry>;
  trust: EnvelopeTrustSummary<EnvelopeSectionForSchema<Version>>;
  provenance: EnvelopeProvenance;
  sensitivity: ClaimSensitivity;
}

export interface EntityIntelligenceEnvelopeV1
  extends EntityIntelligenceEnvelopeBase<typeof ENVELOPE_SCHEMA_VERSION_V1> {
  schemaVersion: typeof ENVELOPE_SCHEMA_VERSION_V1;
  relationships?: never;
}

export interface EntityIntelligenceEnvelopeV2
  extends EntityIntelligenceEnvelopeBase<typeof ENVELOPE_SCHEMA_VERSION_V2> {
  schemaVersion: typeof ENVELOPE_SCHEMA_VERSION_V2;
  relationships?: Paginated<RelationshipsBundle>;
}

export type EntityIntelligenceEnvelope =
  | EntityIntelligenceEnvelopeV1
  | EntityIntelligenceEnvelopeV2;

File: src-tauri/abilities-runtime/src/abilities/get_entity_intelligence/contracts.rs
//! DOS-459 — `EntityIntelligenceEnvelope` DTO contract.
//!
//! Read-side projection over existing claim/proposal/touchpoint substrate.
//! Per L0 packet `.docs/plans/v1.4.4-wp-surface-migration/L0-packet-W1-substrate-gaps.md` §5.1
//! and the locked decisions in §13 (server-signed cursor pagination, per-fact `ProvenanceRef`,
//! typed empty reasons, 1-outer-N-inner block model).
//!
//! Coexists with the existing `get_entity_context` ability — DOS-459 ships the typed
//! envelope; consolidation deferred to v1.5.x per §13 Q1/Q6.

use std::collections::BTreeMap;

use chrono::{DateTime, Utc};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::abilities::list_open_loops::OpenLoop;
use crate::abilities::provenance::SubjectRef;
use crate::abilities::trust::types::TrustBand;
use crate::sensitivity::{ClaimVerificationState, RenderableClaimText};
use crate::types::{ClaimSensitivity, ClaimState, SurfacingState};

/// Schema version for `get_entity_intelligence` envelope.
pub const ENVELOPE_SCHEMA_VERSION: u32 = 1;

// ---- input -----------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum EntityKind {
    Account,
    Project,
    Person,
}

impl EntityKind {
    /// Lower-snake_case identifier matching the legacy `get_entity_context` API.
    pub fn as_lower_str(&self) -> &'static str {
        match self {
            EntityKind::Account => "account",
            EntityKind::Project => "project",
            EntityKind::Person => "person",
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ContextDepth {
    Shallow,
    Standard,
    Deep,
}

impl ContextDepth {
    pub(crate) fn into_legacy(self) -> crate::abilities::get_entity_context::ContextDepth {
        match self {
            ContextDepth::Shallow => crate::abilities::get_entity_context::ContextDepth::Shallow,
            ContextDepth::Standard => crate::abilities::get_entity_context::ContextDepth::Standard,
            ContextDepth::Deep => crate::abilities::get_entity_context::ContextDepth::Deep,
        }
    }
}

/// Envelope section discriminator.
#[derive(Debug, Clone, Copy, Serialize, Deserialize, JsonSchema, PartialEq, Eq, PartialOrd, Ord)]
#[serde(rename_all = "snake_case")]
pub enum EnvelopeSection {
    Facts,
    Health,
    MetadataProposals,
    OpenLoops,
    Touchpoints,
    Threads,
    Record,
}

impl EnvelopeSection {
    /// All section variants in canonical order — `sections` map enumerates these per AC-459.2.
    pub const ALL: &'static [EnvelopeSection] = &[
        EnvelopeSection::Facts,
        EnvelopeSection::Health,
        EnvelopeSection::MetadataProposals,
        EnvelopeSection::OpenLoops,
        EnvelopeSection::Touchpoints,
        EnvelopeSection::Threads,
        EnvelopeSection::Record,
    ];
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct EntityIntelligenceInput {
    pub schema_version: u32,
    pub entity_type: EntityKind,
    pub entity_id: String,
    pub depth: ContextDepth,
    /// `None` = return all sections; non-empty subset = only those sections populated;
    /// excluded sections still appear in `sections` map with `Empty { reason: NotRequested }`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub sections: Option<Vec<EnvelopeSection>>,
}

// ---- cursor + pagination ---------------------------------------------------

/// Opaque server-signed cursor token. Per §13 Q11: cursor signing key + rotation policy
/// is a v1.4.6/v1.4.7 concern (not a substrate change for W1) — at W1 we transport
/// an opaque string and let later substrate sign/verify it.
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(transparent)]
pub struct Cursor(pub String);

impl Cursor {
    pub fn new(token: impl Into<String>) -> Self {
        Self(token.into())
    }

    pub fn as_str(&self) -> &str {
        &self.0
    }
}

/// Per-list cursor lifecycle state (cycle-1 correctness F4).
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum CursorState {
    /// Pagination is consistent — `next_cursor` continues the same logical sequence.
    Stable,
    /// Rows shifted under us (a concurrent insert/retract). Caller MAY continue;
    /// some rows may be skipped or duplicated. Advisory is human-readable.
    DataShifted { advisory: String },
    /// The cursor is no longer valid. Caller MUST restart pagination from page 1.
    Invalidated {
        reason: String,
        restart_required: bool,
    },
}

/// Generic paginated wrapper for envelope list-shape fields.
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct Paginated<T> {
    pub items: Vec<T>,
    pub next_cursor: Option<Cursor>,
    pub total_hint: Option<u64>,
    pub cursor_state: CursorState,
}

impl<T> Paginated<T> {
    pub fn empty_stable() -> Self {
        Self {
            items: Vec::new(),
            next_cursor: None,
            total_hint: Some(0),
            cursor_state: CursorState::Stable,
        }
    }

    pub fn stable(items: Vec<T>) -> Self {
        let total = items.len() as u64;
        Self {
            items,
            next_cursor: None,
            total_hint: Some(total),
            cursor_state: CursorState::Stable,
        }
    }
}

// ---- section state --------------------------------------------------------

/// Typed empty-state reasons per §5.1. No `null` without reason.
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum EmptyReason {
    NotConnected,
    NotProcessedYet,
    FilteredOutBySubject,
    NoRelevantTouchpoints,
    Stale,
    NoEvidenceBackedProposal,
    UnsupportedForSubject,
    NotRequested,
    /// Per AC-459.10 — one section failing emits this; others continue.
    PartialFailure { advisory: String },
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum SectionState {
    Present { item_count: u64 },
    Empty { reason: EmptyReason },
}

// ---- normalized subject ----------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NormalizedSubject {
    pub kind: EntityKind,
    pub id: String,
    pub subject_ref: SubjectRef,
    pub display_label: String,
}

// ---- per-fact provenance reference (ADR-0130 §2 amendment) ----------------

/// Per-fact provenance pointer per cycle-1 architecture F9 — points into the
/// envelope-level `EnvelopeProvenance.sources` index. Prevents 64 KB serialized
/// provenance blowup that ADR-0108 names.
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ProvenanceRef {
    pub source_ids: Vec<String>,
}

impl ProvenanceRef {
    pub fn empty() -> Self {
        Self {
            source_ids: Vec::new(),
        }
    }

    pub fn from_ids(ids: impl IntoIterator<Item = String>) -> Self {
        Self {
            source_ids: ids.into_iter().collect(),
        }
    }
}

// ---- envelope-level provenance + trust + sensitivity ---------------------

/// Display-safe per-source descriptor for the envelope-level provenance index.
/// Raw source identifiers (Glean doc IDs, meeting IDs, URLs, etc.) are redacted
/// per DOS-477 — never emitted unless render policy permits.
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct EnvelopeProvenanceSource {
    /// Stable id used by `ProvenanceRef.source_ids` to point into this index.
    pub id: String,
    pub label: String,
    pub source_type: Option<String>,
    #[schemars(with = "Option<String>")]
    pub as_of: Option<DateTime<Utc>>,
    pub redacted: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct EnvelopeProvenance {
    pub sources: Vec<EnvelopeProvenanceSource>,
    pub redaction_applied: bool,
}

impl EnvelopeProvenance {
    pub fn empty() -> Self {
        Self {
            sources: Vec::new(),
            redaction_applied: false,
        }
    }
}

/// Aggregate trust posture plus per-section caveats.
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct EnvelopeTrustSummary {
    pub aggregate_band: TrustBand,
    pub section_caveats: BTreeMap<EnvelopeSection, String>,
}

impl EnvelopeTrustSummary {
    pub fn unscored() -> Self {
        Self {
            aggregate_band: TrustBand::Unscored,
            section_caveats: BTreeMap::new(),
        }
    }
}

// ---- claim-backed fact, health story, metadata proposals ------------------

/// Claim freshness band — duplicate of receipt enum but lives in the envelope
/// to keep the contract self-describing.
#[derive(Debug, Clone, Copy, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum Freshness {
    Current,
    Aging,
    Stale,
    Unknown,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct EntityFact {
    pub claim_id: String,
    pub subject_ref: SubjectRef,
    pub field_path: Option<String>,
    pub claim_type: String,
    /// Claim text rendered through sensitivity gate — never raw.
    pub rendered_text: RenderableClaimText,
    pub trust_band: TrustBand,
    pub freshness: Freshness,
    #[schemars(with = "Option<String>")]
    pub source_asof: Option<DateTime<Utc>>,
    pub sensitivity: ClaimSensitivity,
    pub lifecycle_state: ClaimState,
    pub surfacing_state: SurfacingState,
    pub verification_state: ClaimVerificationState,
    /// ADR-0130 §2 amendment — per-fact provenance is a *reference* into the envelope's
    /// top-level `EnvelopeProvenance.sources`, NOT an inlined `Vec<ProvenanceSource>`.
    pub provenance: ProvenanceRef,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct HealthStory {
    pub headline: Option<String>,
    pub rows: Vec<HealthStoryRow>,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct HealthStoryRow {
    pub label: String,
    pub body: String,
    pub evidence_claim_ids: Vec<String>,
    pub provenance: ProvenanceRef,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct MetadataProposal {
    pub proposal_id: String,
    pub subject_ref: SubjectRef,
    pub field_path: String,
    pub current_value: Option<String>,
    pub proposed_value: String,
    pub trust_band: TrustBand,
    pub sensitivity: ClaimSensitivity,
    pub provenance: ProvenanceRef,
}

// ---- open loop with receipt target ----------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ReceiptTargetRef {
    pub claim_id: String,
    pub subject_ref: SubjectRef,
    pub field_path: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct OpenLoopWithReceipt {
    pub open_loop: OpenLoop,
    pub receipt_target: ReceiptTargetRef,
    pub trust_band: TrustBand,
    pub freshness: Freshness,
    pub provenance: ProvenanceRef,
}

// ---- touchpoint bundle ----------------------------------------------------

#[derive(Debug, Clone, Copy, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum TouchpointKind {
    Meeting,
    EmailThread,
    Document,
    Salesforce,
    Linear,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum InclusionReason {
    SubjectMatch,
    EntityLink,
    AttendeeMatch,
    DomainMatch,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ExclusionReason {
    SubjectMismatch,
    OutsideWindow,
    LowConfidence,
    Suppressed,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct Touchpoint {
    pub meeting_id: Option<String>,
    pub kind: TouchpointKind,
    #[schemars(with = "String")]
    pub when: DateTime<Utc>,
    pub subject_ref: SubjectRef,
    pub inclusion_reason: InclusionReason,
    pub exclusion_reason: Option<ExclusionReason>,
    pub trust_band: TrustBand,
    pub freshness: Freshness,
    pub provenance: ProvenanceRef,
}

/// Per §5.2 candidate-set primitive — typed basis for "why these touchpoints belong".
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct CandidateSetRef {
    #[schemars(with = "Option<String>")]
    pub window_start: Option<DateTime<Utc>>,
    #[schemars(with = "Option<String>")]
    pub window_end: Option<DateTime<Utc>>,
    pub filter_description: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct SubjectScope {
    pub primary: SubjectRef,
    pub also_includes: Vec<SubjectRef>,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct TouchpointBundle {
    pub upcoming: Paginated<Touchpoint>,
    pub recent: Paginated<Touchpoint>,
    pub candidate_set: CandidateSetRef,
    pub empty_reason: Option<EmptyReason>,
    pub subject_scope: SubjectScope,
}

// ---- threads + record entries --------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ThreadSummary {
    pub thread_id: String,
    pub title: Option<String>,
    #[schemars(with = "Option<String>")]
    pub last_activity_at: Option<DateTime<Utc>>,
    pub message_count: u32,
    pub provenance: ProvenanceRef,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct RecordEntry {
    pub claim_id: String,
    pub subject_ref: SubjectRef,
    pub claim_type: String,
    #[schemars(with = "String")]
    pub recorded_at: DateTime<Utc>,
    pub rendered_text: RenderableClaimText,
    pub trust_band: TrustBand,
    pub sensitivity: ClaimSensitivity,
    pub provenance: ProvenanceRef,
}

// ---- envelope --------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct EntityIntelligenceEnvelope {
    pub schema_version: u32,
    pub subject: NormalizedSubject,
    pub sections: BTreeMap<EnvelopeSection, SectionState>,
    pub facts: Paginated<EntityFact>,
    pub health_story: Option<HealthStory>,
    pub metadata_proposals: Paginated<MetadataProposal>,
    pub open_loops: Paginated<OpenLoopWithReceipt>,
    pub touchpoints: Paginated<TouchpointBundle>,
    pub threads: Paginated<ThreadSummary>,
    pub record_entries: Paginated<RecordEntry>,
    pub trust: EnvelopeTrustSummary,
    pub provenance: EnvelopeProvenance,
    pub sensitivity: ClaimSensitivity,
}
